hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/s2sjoin/SqlStreamToStreamJoinTest.java

```diff
@@ -631,4 +631,30 @@ public void test_joinGenerators() {
                 sqlService.execute("SELECT 1 from TABLE(GENERATE_STREAM(1)) JOIN TABLE(GENERATE_STREAM(3)) on 1=1;"))
                 .hasMessageContaining("For stream-to-stream join, both joined sides must have an order imposed");
     }
+
+    @Test
+    public void test_joinWithImplicitCasts() {
+        String stream = "stream1";
+        TestStreamSqlConnector.create(
+                sqlService,
+                stream,
+                singletonList("a"),
+                singletonList(INTEGER),
+                row(0));
+
+        String stream2 = "stream2";
+        TestStreamSqlConnector.create(
+                sqlService,
+                stream2,
+                singletonList("b"),
+                singletonList(INTEGER),
+                row(0));
+
+        sqlService.execute("CREATE VIEW s1 AS SELECT * FROM TABLE(IMPOSE_ORDER(TABLE stream1, DESCRIPTOR(a), 1))");
+        sqlService.execute("CREATE VIEW s2 AS SELECT * FROM TABLE(IMPOSE_ORDER(TABLE stream2, DESCRIPTOR(b), 1))");
+
+        assertTipOfStream(
+                "SELECT * FROM s1 JOIN s2 ON s2.b BETWEEN s1.a AND s1.a + 1",
```

Comment:
`"SELECT * FROM s1 JOIN s2 ON s2.b BETWEEN s1.a AND CAST(CAST(s1.a as VARCHAR) AS TIMESTAMP) + INTERVAL '10' SECONDS"` fails during execution (`Cannot convert INTEGER to TIMESTAMP`)

Reply:
should it be possible to mix `int` and `timestamp` watermarks? `SELECT * FROM s1 JOIN s2 ON s2.b BETWEEN s1.a AND s1.a + 1` in updated test fails during execution. We currently do not have `INTEGER` <-> `TIMESTAMP` conversions

Reply:
enough manual fuzzing for now

Reply:
1. Interval type cannot be assigned to column, it's just an argument type for TIMESTAMP(_WITH_TZ) type to add/subtract constants. More info : https://github.com/hazelcast/hazelcast/pull/18390.
2. According to type coercion table, `INT <-> TIMESTAMP` conversion is not allowed. You need to use `TO_TIMESTAMP_TZ` function for that.
3.  > should it be possible to mix int and timestamp watermarks? 

Only with `TO_TIMESTAMP_TZ` function usage. Overall, S2S JOIN feature was designed with the understanding that ~99% of watermarked fields are timestamps (as a truly monotonically increasing value), but we also allowed integers to be watermarked too.

Reply:
> I get NPE for query "SELECT * FROM s1 JOIN s2 ON s2.b BETWEEN s1.a AND CAST(s1.a as INTERVAL DAY TO SECONDS) + INTERVAL '10' SECONDS"

Its an action item for another issue, this patch doesn't intend to fix NPE for usage of INTERVAL type assignment.

Reply:
ad.1. Should produce better error
ad.2. Should fail during query validation not in Jet processor
ad.3. This will not work because we do not support `TO_TIMESTAMP_TZ` in `addAddends` but this can be implemented separately if needed. Thx for explanation.

Without CAST support for windows we probably rejected such malformed queries earlier. Now we need to validate them more thoroughly.

Reply:
very rudimentary CAST implementation (skipping it) allows some assumptions to fail later